### PR TITLE
[Contributing][Code] do not distinguish regular classes and API classes

### PR DESCRIPTION
**Pull request by xabbuh**
| Q | A |
| --- | --- |
| Doc fix? | no |
| New docs? | kind of (symfony/symfony#15977, symfony/symfony#15979) |
| Applies to | all |
| Fixed tickets |  |

### CODE REVIEW

**Comment by stof** [comment]
I would just add a small note saying that these BC rules don't apply on classes/methods tagged as ``@internal``

**Comment by xabbuh** [comment]
@stof We already have this caution blocks (and a similar one for classes):

>  The exception to this rule are interfaces tagged with ``@internal``. Such interfaces should not be used or implemented.

**Comment by stof** [comment]
Then fine.

**Comment by WouterJ** [comment]
:+1:

File: contributing/code/bc.rst [hunk 1]
```diff
@@ -51,53 +51,28 @@ sticks to these rules.
...
-| **If you implement the interface and...**     | **Then we guarantee BC...**   |
-+-----------------------------------------------+---------------+---------------+
-| Implement a method                            | No [1]_       | Yes           |
-+-----------------------------------------------+---------------+---------------+
-| Add an argument to an implemented method      | No [1]_       | Yes           |
-+-----------------------------------------------+---------------+---------------+
-| Add a default value to an argument            | Yes           | Yes           |
-+-----------------------------------------------+---------------+---------------+
++-----------------------------------------------+-----------------------------+
+| Use Case                                      | Backwards Compatibility     |
```
Comment (javiereguiluz) [inline contributing/code/bc.rst]: We should rethink this table. In the past some columns were `Yes` and others `No`, so the table made sense. But now everything is `Yes`.

Comment (WouterJ) [inline contributing/code/bc.rst]: I think a table structure is very nice here. It's clear and it's consistent with the other "promises tables" below.

**Comment by weaverryan** [comment]
Nice work Christian - thanks!

File: contributing/code/bc.rst [hunk 2]
```diff
@@ -234,134 +177,96 @@ Changing Classes
...
-Remove type hint of an argument                     Yes [2]_        No
-Change argument type                                Yes [2]_ [5]_   No
-Change return type                                  Yes [2]_ [6]_   No
+Add public method                                   Yes
+Remove public method                                No
+Change name                                         No
+Reduce visibility                                   No
+Move to parent class                                Yes
+Add argument without a default value                No
+Add argument with a default value                   No
```
Comment (webmozart) [inline contributing/code/bc.rst]: This was a brave change. Now it's impossible to extend existing public code with new functionality. Was this intended?

see https://github.com/symfony/symfony/pull/16834

Comment (xabbuh) [inline contributing/code/bc.rst]: This was the only option we had as PHP will complain anyway (see https://3v4l.org/6T2DO).

**Pull request merged**